src/inputfetcher.h

```diff
@@ -0,0 +1,197 @@
+ */
+class InputFetcher
+{
+private:
+    //! The latest input being fetched. Workers atomically increment this when fetching.
+    alignas(64) std::atomic_size_t m_input_head{0};
+
+    //! The inputs of the block which is being fetched.
+    struct Input {
+        enum class Status : uint8_t {
+            WAITING, // The coin has not been fetched yet
+            READY, // The coin has been fetched and is ready to be inserted into the cache
+            FAILED, // The coin failed to be fetched
+            SKIPPED, // The coin is created and spent in the same block so cannot be fetched
+        };
+
+        //! Workers update this after setting the coin. The main thread spins on this until it is not WAITING.
+        std::atomic<Status> status{Status::WAITING};
+        //! The outpoint to fetch;
+        const COutPoint& outpoint;
+        //! The coin that workers will fetch and main thread will insert into cache.
+        Coin coin{};
+
+        Input(Input&& other) noexcept : outpoint{other.outpoint} {} // Only moved in setup for reallocation.
+        Input(const COutPoint& o LIFETIMEBOUND) noexcept : outpoint{o} {}
+    };
+    std::vector<Input> m_inputs{};
+
+    /**
+     * The set of txids of all txs in the block being fetched.
+     * Used to filter out inputs that are created and spent in the same block,
+     * since they will not be in the db or the cache.
+     */
+    std::unordered_set<Txid, SaltedTxidHasher> m_txids{};
+
+    //! DB coins view to fetch from.
+    const CCoinsView* m_db{nullptr};
+    //! The cache to fetch from.
+    const CCoinsViewCache* m_cache{nullptr};
+
+    std::vector<std::thread> m_worker_threads{};
+    std::barrier<> m_barrier;
+    bool m_request_stop{false};
+
+    void WorkLoop() noexcept
+    {
+        while (true) {
+            m_barrier.arrive_and_wait();
+            if (m_request_stop) [[unlikely]] {
+                return;
+            }
+            while (true) {
+                const size_t i{m_input_head.fetch_add(1, std::memory_order_relaxed)};
+                if (i >= m_inputs.size()) [[unlikely]] {
+                    break;
+                }
+                auto& input{m_inputs[i]};
+                // If an input spends an outpoint from earlier in the block,
+                // it won't be in the cache yet but it also won't be in the db either.
+                if (m_txids.contains(input.outpoint.hash)) {
+                    input.status.store(Input::Status::SKIPPED, std::memory_order_relaxed);
+                    continue;
+                }
+                try {
+                    if (auto coin{m_cache->GetPossiblySpentCoinFromCache(input.outpoint)}) {
+                        input.coin = std::move(*coin);
+                        if (!input.coin.IsSpent()) [[likely]] { // Coin from cache could be spent
+                            // We need release here, so setting coin 2 lines above happens before the main thread loads.
```

Comment:
The comments indicate that you also don't think this code is very intuitive - I'm also a bit lost here, can we simplify this somehow? I don't even understand what "release" means here or why both branches result in `Status::READY` (can we unify them) or what happens if the first internal if isn't fulfilled, or the outer one or the else, etc. The branching + `continue` + try/catch doesn't help

Reply:
Rewrote this part to make it more clear. I was trying to be clever by avoiding some extra checks, but they are probably meaningless and clarity is better here.